File: 002_Systematic_Trading/002-Chan_Jegadeesh_Lakonishok-Momentum_Strategies-1996/code/object_oriented/python_3_11/src/momentum_calculator.py
from typing import List, Dict, TypedDict
from dataclasses import dataclass
import csv
import math
from collections import defaultdict

class EpsRecord(TypedDict):
    company_id: str
    quarter: str
    eps: float

@dataclass
class EpsData:
    company_id: str
    quarter: str
    eps: float

    @staticmethod
    def from_dict(record: EpsRecord) -> 'EpsData':
        return EpsData(record['company_id'], record['quarter'], float(record['eps']))

class MomentumCalculator:
    def __init__(self, lookback: int = 4):
        self.lookback = lookback

    def compute_signal(self, eps_data: List[EpsData]) -> List[tuple[str, float]]:
        # Group by company
        company_eps: Dict[str, List[EpsData]] = defaultdict(list)
        for record in eps_data:
            company_eps[record.company_id].append(record)

        results = []
        for company_id, records in company_eps.items():
            # Sort by quarter (assuming Q1_YYYY format)
            records.sort(key=lambda x: self._parse_quarter(x.quarter))
            if len(records) < self.lookback + 1:
                continue

            # Compute momentum for the latest quarter
            latest_eps = records[-1].eps
            past_eps = records[-1 - self.lookback].eps
            eps_values = [r.eps for r in records[-5:]]  # Last 5 quarters for stddev
            stddev = self._stddev(eps_values)
            if stddev == 0:
                continue
            momentum = (latest_eps - past_eps) / stddev
            results.append((company_id, momentum))

        return results

    def _parse_quarter(self, quarter: str) -> tuple[int, int]:
        match quarter.split('_'):
            case ['Q1', year]:
                return (int(year), 1)
            case ['Q2', year]:
                return (int(year), 2)
            case ['Q3', year]:
                return (int(year), 3)
            case ['Q4', year]:
                return (int(year), 4)
            case _:
                raise ValueError(f"Invalid quarter format: {quarter}")

    def _stddev(self, values: List[float]) -> float:
        if not values:
            return 0.0
        mean = sum(values) / len(values)
        variance = sum((x - mean) ** 2 for x in values) / len(values)
        return math.sqrt(variance)

def load_eps_data(file_path: str) -> List[EpsData]:
    with open(file_path, 'r') as f:
        reader = csv.DictReader(f)
        return [EpsData.from_dict(row) for row in reader]

if __name__ == "__main__":
    calculator = MomentumCalculator()
    eps_data = load_eps_data("eps_data.csv")
    signals = calculator.compute_signal(eps_data)
    for company_id, signal in signals:
        print(f"{company_id}: {signal:.4f}")

File: 002_Systematic_Trading/002-Chan_Jegadeesh_Lakonishok-Momentum_Strategies-1996/code/object_oriented/cpp_26/src/momentum_calculator.cpp
#include <algorithm>
#include <cmath>
#include <fstream>
#include <iostream>
#include <ranges>
#include <sstream>
#include <string>
#include <vector>
#include <flat_map>
#include <format>

struct EpsData
{
    std::string company_id;
    std::string quarter;
    double eps;

    static EpsData from_csv_line(const std::string &line)
    {
        std::stringstream ss(line);
        std::string company_id, quarter, eps_str;
        std::getline(ss, company_id, ',');
        std::getline(ss, quarter, ',');
        std::getline(ss, eps_str, ',');
        return {company_id, quarter, std::stod(eps_str)};
    }
};

class MomentumCalculator
{
public:
    explicit MomentumCalculator(int lookback = 4) : lookback_(lookback) {}

    auto compute_signal(const std::vector<EpsData> &eps_data) const
    {
        // Group by company using ranges
        std::flat_map<std::string, std::vector<EpsData>> company_eps;
        for (const auto &record : eps_data)
        {
            company_eps[record.company_id].push_back(record);
        }

        std::vector<std::pair<std::string, double>> results;
        for (auto &[company_id, records] : company_eps)
        {
            auto sorted_records = records | std::ranges::to<std::vector>();
            std::ranges::sort(sorted_records, {}, &EpsData::quarter);
            if (sorted_records.size() < static_cast<size_t>(lookback_ + 1))
            {
                continue;
            }

            const double latest_eps = sorted_records.back().eps;
            const double past_eps = sorted_records[sorted_records.size() - 1 - lookback_].eps;
            auto eps_values = sorted_records | std::views::take_last(5) | std::views::transform(&EpsData::eps) | std::ranges::to<std::vector>();
            const double sigma = stddev(eps_values);
            if (sigma == 0.0)
            {
                continue;
            }
            results.emplace_back(company_id, (latest_eps - past_eps) / sigma);
        }
        return results;
    }

private:
    int lookback_;

    static double stddev(const std::vector<double> &values)
    {
        if (values.empty())
            return 0.0;
        const double mean = std::ranges::sum(values) / values.size();
        const double variance = std::ranges::fold_left(
                                    values | std::views::transform([&](double x)
                                                                   { return (x - mean) * (x - mean); }),
                                    0.0,
                                    std::plus{}) /
                                values.size();
        return std::sqrt(variance);
    }
};

std::vector<EpsData> load_eps_data(const std::string &file_path)
{
    std::ifstream file(file_path);
    std::vector<EpsData> data;
    std::string line;
    std::getline(file, line); // Skip header
    while (std::getline(file, line))
    {
        data.push_back(EpsData::from_csv_line(line));
    }
    return data;
}

int main()
{
    MomentumCalculator calc;
    auto eps_data = load_eps_data("eps_data.csv");
    auto signals = calc.compute_signal(eps_data);
    for (const auto &[company_id, signal] : signals)
    {
        std::cout << std::format("{}: {:.4f}\n", company_id, signal);
    }
    return 0;
}
